### DISCUSSION TITLE
[lodash] - Property 'find' does not exist on type 'LoDashExplicitWrapper<any>'

### DISCUSSION THREAD

**Discussion by jahuja-coalition** [DefinitelyTyped/DefinitelyTyped#61531] [Issues with a @types package]
I'm wondering why code like this:

```
 return _.chain(requestBody)
    .get(["request", "custom_fields"], [])
    .find({ "id": currentId })
``` 

would trigger the following error:

```
Property 'find' does not exist on type 'LoDashExplicitWrapper<any>'
```

We're seeing this after upgrading lodash types from `4.14.112` to `4.14.182`.

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
This should work as long as TypeScript selects the correct overload (should be https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/object.d.ts#L1200). What is the type of `requestBody` in this context? If it's `any` or `unknown` then it may not select the correct overload - in this case it looks like it's using the last overload which is wrong.

(More context: a while ago the chain wrapper types were overhauled to improve performance. The old system was more powerful but unfortunately caused horribly slow compile times in many cases. The new system is much faster but might not handle all of the edge cases correctly. I suspect that is why you are seeing an error after upgrading.)

**Reply by jahuja-coalition** [reply]
Thanks for your response @aj-r! The type of requestBody isn't `any` or `unknown`, it looks like this:

```
export interface TicketField { id: number; value: number | string; }
export interface TicketResponse {
  request: {
    url: string;
    id: number;
    status: "new" | "open" | "pending" | "hold" | "solved" | "closed";
    priority: string | null;
    type: string | null;
    subject: string;
    description: string;
    organization_id: number | null;
    via: {
      channel: string;
      source: {
        from: object;
        to: object;
        rel: null;
      };
    };
  };
  custom_fields: TicketField[];
  requester_id: number;
  collaborator_ids: number[];
  email_cc_ids: number[];
  is_public: boolean;
  due_at: null | string;
  can_be_solved_by_me: boolean;
  created_at: string;
  updated_at: string;
  recipient: string | null;
  followup_source_id: number | null;
  assignee_id: number | null;
  ticket_form_id: number;
  fields: TicketField[];
}
```

Do you have any suggested workarounds for the `.find` error, or would we need to wait for this edge case to be handled?

**Reply by jahuja-coalition** [reply]
In addition, we also see errors like this:

```
error TS2339: Property 'email' does not exist on type 'object'.
```

on code like this:

```
  if (_.isEmpty(req.body.email)) {
    return res.status(HttpStatus.BAD_REQUEST).json({
      "message": "Missing required POST request body parameter: email"
    });
  }
```

Any thoughts there?

**Reply by aj-r** [reply]
In that case I"m not sure what is going on. If any field is possibly null/undefined (or the requestBody itself could be null/undefined) then maybe that overload won't work as expected. It might help to change the `get` overloads to use `NonNullable<>` in several places to account for this, but this could add a large amount of complexity.

A simpler alternative is you could avoid using `get` altogether. Now that TypeScript has the `?.` and `??` operators, there is rarely any reason to use `get` anymore. e.g.

```ts
const customFields = requestBody?.request?.custom_fields ?? [];
return _.find(customFields, { "id": currentId });
```

Regarding the other error (`Property 'email' does not exist on type 'object'`), it looks like that error is unrelated to lodash. I'm not sure what `req.body` is, but maybe it isn't typed correctly.
